### DISCUSSION TITLE
Tokenization of decimals/periods

### DISCUSSION THREAD

**Discussion by raqibhayder** [explosion/spaCy#6935] [Help: Coding & Implementations]
I am trying to detect phone numbers in a given piece of text. I have added a few basic examples below.  

``` python
import spacy

nlp = spacy.load("en_core_web_sm")

phone_numbers = ["555-435-2234", 
                 "555 555.888", 
                 "555 555-5555", 
                 "556 456 4785", 
                 "789.326.2563", 
                 "776.444 4569"]

for i, phone_number in enumerate(phone_numbers): 
  doc = nlp(phone_number)
  print(f"Example {i+1}:")
  print(f"Text: {doc.text}")
  print(f"LIKE_NUM: {[token.like_num for token in doc]}")
  print(f"Tokens: {nlp.tokenizer.explain(phone_number)}")
  print(f"Entities: {[(ent.label_, ent.text) for ent in doc.ents]} \n")
  
```

---

**Questions:**

**Example 1, 3**: I understand how the tokenization works for this, but do not understand why is the entity labelled as `CARDINAL`? What makes an entity a `CARDINAL`?

**Example 2:** How do I make spaCy treat the `.` as a period and not a decimal? In the context that I am working in, almost all decimals are supposed to be treated as periods? How do I go about doing that?

**Example 4**: Why are the first two tokens treated as `CARDINAL`? What about the third token?

**Example 5**: Why is this a single token? Why does spaCy think this is `like_num`?

**Example 6**: What is the difference between 4 and 6? 

---

**Output**

``` 
Example 1:
Text: 555-435-2234
LIKE_NUM: [True, False, True, False, True]
Tokens: [('TOKEN', '555'), ('INFIX', '-'), ('TOKEN', '435'), ('INFIX', '-'), ('TOKEN', '2234')]
Entities: [('CARDINAL', '555-435-2234')] 

Example 2:
Text: 555 555.888
LIKE_NUM: [True, True]
Tokens: [('TOKEN', '555'), ('TOKEN', '555.888')]
Entities: [] 

Example 3:
Text: 555 555-5555
LIKE_NUM: [True, True, False, True]
Tokens: [('TOKEN', '555'), ('TOKEN', '555'), ('INFIX', '-'), ('TOKEN', '5555')]
Entities: [('CARDINAL', '555 555-5555')] 

Example 4:
Text: 556 456 4785
LIKE_NUM: [True, True, True]
Tokens: [('TOKEN', '556'), ('TOKEN', '456'), ('TOKEN', '4785')]
Entities: [('CARDINAL', '556 456')] 

Example 5:
Text: 789.326.2563
LIKE_NUM: [True]
Tokens: [('TOKEN', '789.326.2563')]
Entities: [('CARDINAL', '789.326.2563')] 

Example 6:
Text: 776.444 4569
LIKE_NUM: [True, True]
Tokens: [('TOKEN', '776.444'), ('TOKEN', '4569')]
Entities: [('CARDINAL', '776.444 4569')] 
```

I used the code above to understand how spaCy is tokenizing and detecting entities. My ultimate goal is to detect phone numbers, and be able to use a pattern like below. This is just one pattern, I am working on detecting most NANP, and European phone numbers from texts. Will be building on top of this. 

``` python
import spacy
from spacy.pipeline import EntityRuler

nlp = spacy.load("en_core_web_sm")

phone_numbers = ["555-435-2234", 
                 "555 555.888", 
                 "555 555-5555", 
                 "556 456 4785", 
                 "789.326.2563", 
                 "776.444 4569"]

entity_ruler = EntityRuler(nlp, overwrite_ents=True)
phone_pattern = [{"label": "PHONE", 
                  "pattern": [{"SHAPE": "ddd"}, 
                              {"TEXT": {"REGEX": r"[.-]"}, "OP": "?"}, 
                              {"SHAPE": "ddd"}, 
                              {"TEXT": {"REGEX": r"[.-]"}, "OP": "?"}, 
                              {"SHAPE": "dddd"}]
                  }]
entity_ruler.add_patterns(phone_pattern)
nlp.add_pipe(entity_ruler, name="phone", before="ner")

for i, phone_number in enumerate(phone_numbers): 
  doc = nlp(phone_number)
  print(f"Example {i+1}:")
  print(f"Text: {doc.text}")
  print(f"LIKE_NUM: {[token.like_num for token in doc]}")
  print(f"Tokens: {nlp.tokenizer.explain(phone_number)}")
  print(f"Entities: {[(ent.label_, ent.text) for ent in doc.ents]} \n")
```

**Output**

```
Example 1:
Text: 555-435-2234
LIKE_NUM: [True, False, True, False, True]
Tokens: [('TOKEN', '555'), ('INFIX', '-'), ('TOKEN', '435'), ('INFIX', '-'), ('TOKEN', '2234')]
Entities: [('PHONE', '555-435-2234')] 

Example 2:
Text: 555 555.888
LIKE_NUM: [True, True]
Tokens: [('TOKEN', '555'), ('TOKEN', '555.888')]
Entities: [] 

Example 3:
Text: 555 555-5555
LIKE_NUM: [True, True, False, True]
Tokens: [('TOKEN', '555'), ('TOKEN', '555'), ('INFIX', '-'), ('TOKEN', '5555')]
Entities: [('PHONE', '555 555-5555')] 

Example 4:
Text: 556 456 4785
LIKE_NUM: [True, True, True]
Tokens: [('TOKEN', '556'), ('TOKEN', '456'), ('TOKEN', '4785')]
Entities: [('PHONE', '556 456 4785')] 

Example 5:
Text: 789.326.2563
LIKE_NUM: [True]
Tokens: [('TOKEN', '789.326.2563')]
Entities: [('CARDINAL', '789.326.2563')] 

Example 6:
Text: 776.444 4569
LIKE_NUM: [True, True]
Tokens: [('TOKEN', '776.444'), ('TOKEN', '4569')]
Entities: [('CARDINAL', '776.444 4569')] 
```

What are your views on my approach? 

I am also experimenting using regex directly to detect phone numbers on `doc.text`, and assigning entity labels using `doc.char_span` in a custom pipeline component.

What would be the recommended approach?

Thank you 🙏

**Comment by gkowalski** [comment]
Im sure your answer is buried in https://spacy.io/usage/linguistic-features#how-tokenizer-works I need to figure this out for dates , which have the same separator issues

**Comment by gkowalski** [comment]
I am running code like : 
```            
infix_re = re.compile(r'''[-/,:;]''')
suffix_re = re.compile(r'''["']''')  # Handle suffix chars
nlp.tokenizer = custom_tokenizer(nlp, infix_re, suffix_re)

ruler = nlp.add_pipe("entity_ruler", config={"validate":True , "overwrite_ents":True}).from_disk("../spacy_patterns.jsonl")
```

and mapping my own entities to "MDATE" over DATE   (here is part of my spacy_patterens.jsonl ) : 

```
{"label":"MDATE","example":"MM/DD/YYYY and YYYY/MM/DD","pattern":[{"IS_DIGIT":true},{"ORTH":"/"},{"IS_DIGIT":true},{"ORTH":"/"},{"IS_DIGIT":true}]}
{"label":"MDATE","example":"MM-DD-YYYY and YYYY-MM-DD","pattern":[{"IS_DIGIT":true},{"ORTH":"-"},{"IS_DIGIT":true},{"ORTH":"-"},{"IS_DIGIT":true}]}
{"label":"MDATE","example":"MM.DD.YYYY and YYYY.MM.DD","pattern":[{"IS_DIGIT":true},{"ORTH":"."},{"IS_DIGIT":true},{"ORTH":"."},{"IS_DIGIT":true}]}
```

The period version  is still now working yet.

**Comment by polm** [comment]
Sorry for the late reply. I think we missed this one because you have a lot of detailed questions here, and it's kind of a lot of work to answer this. 

Some general notes:

- `CARDINAL` is defined in spaCy as "Numerals that do not fall under another type", so basically there's no deep reason to it.
- To understand why the tokenizer did a thing, use the [explainer](https://spacy.io/api/tokenizer#explain).
- If you have to dig into the tokenizer deeply, you should either work around it or dig into the source.

For dates and phone numbers, if you care about handling irregular formatting, it is probably possible to get the tokenizer to do what you want but also probably not worth the effort (especially since you don't want to change tokenization elsewhere in the sentence). I would just add a component that runs regexes against the raw doc text and use [Doc.char_span](https://spacy.io/api/doc#char_span) to add entities. If you put that before any other EntityRuler/NER components the annotations won't be overwritten.
